spring-integration-core/src/main/java/org/springframework/integration/mapping/AbstractHeaderMapper.java

```diff
@@ -151,7 +152,22 @@ else if (NON_STANDARD_HEADER_NAME_PATTERN.equals(pattern)) {
 				matchers.add(new PrefixBasedMatcher(false, this.standardHeaderPrefix));
 			}
 			else {
-				matchers.add(new PatternBasedHeaderMatcher(Collections.singleton(pattern)));
+				String thePattern = pattern;
+				boolean negate = false;
+				if (pattern.startsWith("!")) {
+					thePattern = pattern.substring(1);
+					negate = true;
+				}
+				else if (pattern.startsWith("\\!")) {
+					thePattern = pattern.substring(1);
+				}
+				if (negate) {
+					// negative matchers get priority
+					matchers.add(0, new PatternBasedHeaderMatcher(Collections.singleton(thePattern), negate));
+				}
+				else {
+					matchers.add(new PatternBasedHeaderMatcher(Collections.singleton(thePattern), negate));
```

Comment:
Yes; I thought it a bit odd that each one only has one pattern; will revise on Tuesday.